exp/controllers/helpers.go

```diff
@@ -408,11 +408,28 @@ func MachinePoolToAzureManagedControlPlaneMapFunc(ctx context.Context, c client.
 			}
 		}
 
-		nameMatches := controlPlane.Spec.DefaultPoolRef.Name == infraMachinePoolRef.Name
 		kindMatches := infraMachinePoolRef.Kind == "AzureManagedMachinePool"
 		groupMatches := controlPlaneGK.Group == gv.Group
 
-		if groupMatches && kindMatches && nameMatches {
+		ammp := &infrav1exp.AzureManagedMachinePool{}
+		key := types.NamespacedName{Namespace: infraMachinePoolRef.Namespace, Name: infraMachinePoolRef.Name}
+		if err := c.Get(ctx, key, ammp); err != nil {
+			log.Error(err, "failed to fetch azure managed machine pool for Machinepool: %s", infraMachinePoolRef.Name)
+			// If we get here, we might want to reconcile but aren't sure.
+			// Do it anyway to be safe. Worst case we reconcile a few extra times with no-ops.
+			return []reconcile.Request{
+				{
+					NamespacedName: client.ObjectKey{
+						Namespace: ref.Namespace,
+						Name:      ref.Name,
+					},
+				},
+			}
+		}
+
+		isSystemNodePool := ammp.Spec.Mode == infrav1exp.SystemNodePool
+
+		if groupMatches && kindMatches && isSystemNodePool {
```

Comment:
hmm...I'm not 100% sure I follow the thinking here. when this was the default pool, we reconciled it as part of the managedcluster itself (actually the managed service still tries to do this for all pools, maybe a TODO there). we'd actually double reconcile it I think, once as part of MC and once as part of AP.

if we modified the last system pool however, we don't want to reconcile the cluster, we want to reconcile just that pool. if we accidentally race and try to delete the last system pool before creating a new one, just let the retry happen on 400 (return err from reconcile)

we can probably also make the managedclusters client avoid patching agentpools after create, to cleanly separate that

Reply:
MachinePoolToAzureManagedControlPlaneMapFunc was added by you in #1397 
This is the description, reasoning the PR has.
`If the MachinePool owning the defaultPool of an AzureManagedControlPlane is nil when we try to reconcile, we need to re-reconcile. We can either return an error or watch machinePools and filter them down to the one owning the defaultPool (more accurate, but requires watching and filtering all machinePool events). I implemented the latter.
`

So this is used in AMCP and not in MC.  

I found your reasoning fine hence i thought of working on the same lines for system node pool, i.e if the owner mp of system node pool is nil we try to reconcile as it is more accurate. 

But I guess I made a mistake in line 265 in azuremanagedcontrolplane_reconciler I should return a nil over there.

Reply:
Yeah, I see. This makes sense, but then we should definitely make sure AMCP reconciler doesn't send all the AMMPs to AKS on updates, only update the control plane config. We don't want AMCP and AMMP reconcilers to compete to reconcile node pools *except* that we let AMCP create the very first one as part of the cluster.

Reply:
yeah makes sense, i think we have a check for this in line 216 azuremanagedcontrolplane.go. In amcp we reconcile ammp only during creation.